### DISCUSSION TITLE
Are shortest path algorithms available for MultiDiGraph?

### DISCUSSION THREAD

**Discussion by dg-pb** [networkx/networkx#7000] [Q&A] [answered]
It seems that algorithm works, but the output does not contain number specifying which connection between 2 nodes is being referred to.

Am I doing something wrong?

```python
import networkx as nx
G = nx.MultiDiGraph()
G.add_edges_from([('a', 'b'), ('b', 'c'), ('b', 'c')], weight=1)
print(nx.shortest_path(G, weight='weight'))
```

RESULT
```
{'a': {'a': ['a'], 'b': ['a', 'b'], 'c': ['a', 'b', 'c']},
 'b': {'b': ['b'], 'c': ['b', 'c']},
 'c': {'c': ['c']}}
```

**Comment by Schefflera-Arboricola** [comment]
[SEE THIS COMMENT](https://github.com/networkx/networkx/discussions/7000#discussioncomment-7242684)

[EDIT] @dgrigonis In `_weight_function(G, weight)` only the edge weights were extracted, and for MultiGraph the edge with minimum weight was taken. ~~The edge attributes and key info used to get lost for MultiGraphs. But, this was fixed in this commit: https://github.com/networkx/networkx/commit/adbaaa1a0e9d7e94d3623d0fe52157627af1228e~~

~~I think it will be updated in the next release!~~

Really sorry for the misleading comment. I'd be more careful next time.

**Reply by dg-pb** [reply]
Thank you, will look forward

**Reply by dschult** [reply, marked as answer]
The comment above is a little misleading. That commit fixed a different problem (negative edge cycles).
The NetworkX shortest path functions act on multigraphs and return a list of nodes indicating the path. The user must find which edge is used by tracing the node-pairs in the path and choosing the desired edge. Often the desired edge is the edge between the two nodes that has minimal weight, but that depends on your `weight` function and how you handle multiedges when reporting weights.  A helpful networkx utility for looping over the node-pairs in a path is `nx.utils.pairwise(path)` as in:
```python
for u, v in nx.utils.pairwise(path):
    ekey_dict = G[u][v]
    ....
```

**Reply by Schefflera-Arboricola** [reply]
Sorry for the inconvenience caused and thank you for pointing it out @dschult sir.

**Reply by dg-pb** [reply]
Thank you, all makes sense and working.
